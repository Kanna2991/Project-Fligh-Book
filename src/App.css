.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
}
.App-header .parent{
  display: grid;
  grid-gap: 15px;
}
.App-header .child{
  display: flex;
}
.App-header .child > div{
  width: 30px;
  height: 50px;
  border: 1px solid black;
  pointer-events: none;
  background-color: lightgray;
  color: white;
  display: grid;
  place-content: center;
}
.App-header .child > div.available{
  background-color: lightgreen;
  pointer-events: all;
  cursor: pointer;
}
.App-header .character {
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  margin: 25px 0;
  width: 50%;
}
.App-header .character div span {
  display: inline-block;
  padding: 8px;
  background-color: lightgreen;
}
.App-header .character div span.booked, .App-header .child > div.booked {
  background-color: red;
}
.App-header .character div span.yetToBook {
  background-color: lightgray;
}
/* .App-header .child .window{
  background-color: green;
} 
.App-header .child .center{
  background-color: red;
} 
.App-header .child .aisle{
  background-color: blue;
} */
.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
