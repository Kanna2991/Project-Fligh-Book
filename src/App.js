import React, { useCallback, useEffect, useState } from "react";
import './App.css';

function App({initialRowsColumns}) {
    const [resultArray, setNewResult] = useState(null);
    const [passengerNo, setPassengerNo] = useState(0);
    const generateWorkingArray = useCallback(() => {
        const input2DArray = initialRowsColumns || JSON.parse(localStorage.getItem('input2DArray')) || [ [3,2], [4,3], [2,3], [3,4] ];
        const resultArray = input2DArray.map((childArray, i) => {
            const tempArray = [];
            for(let j = 0; j < Number(childArray[1]); j++) {
                const arrayObj = [];
                for(let a = 0; a < Number(childArray[0]); a++) {
                    let seatType = (i === 0 && a === 0) || ((i === (input2DArray.length - 1)) && (a === (childArray[0]-1))) ? 'window'
                        : (a !== 0 && a !== (Number(childArray[0]) - 1) ? 'center' : 'aisle');
                    let available = (input2DArray.length === 1) && (j+1 === 1) && (a === 0) && (seatType === 'window');
                    available = available || ((j+1 === 1) && (i+1 === 1) && (seatType === 'aisle'));
                    arrayObj.push({id: i + '' + j + '' + a, seatType, row: j+1, parentColumn: i+1, available});
                }
                tempArray.push(arrayObj);
            }
            return tempArray;
        });
        setNewResult(resultArray);
    }, [initialRowsColumns]);

    useEffect(()=>{
        generateWorkingArray();
    }, [generateWorkingArray]);

    const computeColumn = (arr) => {
        let templateColumn = '';
        arr?.forEach(item => { templateColumn += 'auto ';});
        return templateColumn;
    }

    const findSeatExist = (arr, seat) => {
        let seatFound = null;
        let filterSeatType = []
        arr.forEach(arr1 => {
            arr1.forEach(arr2 => {
                arr2.forEach(item => {
                    if(!seatFound && seat.seatType === item.seatType && seat.id !== item.id && !item.booked) {
                        seatFound = item;
                    }
                    if(seat.seatType === item.seatType && seat.id !== item.id && !item.booked) {
                        filterSeatType.push(item);
                    }
                })
            })
        });
        let sameRowExist = filterSeatType.find(item => item?.row === seat?.row)
        return sameRowExist || seatFound;
    }

    const setExistArray = (bookedSeatExist, tempArray) => {
        const existIds = (bookedSeatExist.id).split('').map(i => Number(i));
        tempArray[existIds[0]][existIds[1]][existIds[2]] = {...bookedSeatExist, available: true};
        setNewResult(tempArray);
    }
    const openNewAvailableSeat = (arr, bookedSeat) => {
        const tempArray = JSON.parse(JSON.stringify(arr));
        const ids = (bookedSeat.id).split('').map(i => Number(i));
        const bookedChildRowArr = tempArray[ids[0]][ids[1]];
        const sameRowAvailExist = bookedChildRowArr.findIndex(item => bookedSeat.id !== item.id && item.seatType === bookedSeat.seatType && !item.booked)
        if(sameRowAvailExist > -1) {
            tempArray[ids[0]][ids[1]][sameRowAvailExist] = {...tempArray[ids[0]][ids[1]][sameRowAvailExist], available: true};
            setNewResult(tempArray);
            return;
        }
        if(!(tempArray[ids[0] + 1])){
            let parentIndex = tempArray.length === ids[0] + 1 ? 0 : ids[0];
            let childIndex = tempArray[parentIndex].length === ids[1] + 1 || tempArray[parentIndex].length < ids[1] + 1 ? 0 : (ids[1] + 1);
            let availIndex = tempArray[parentIndex][childIndex].findIndex(item => item.seatType === 'aisle' && !item.booked);
            availIndex = availIndex > -1 ? availIndex : tempArray[parentIndex][childIndex].findIndex(item => item.seatType === 'window' && !item.booked);
            availIndex = availIndex > -1 ? availIndex : tempArray[parentIndex][childIndex].findIndex(item => item.seatType === 'center' && !item.booked);
            if(!tempArray[parentIndex][childIndex][availIndex] || (tempArray[parentIndex][childIndex][availIndex]['seatType'] !== bookedSeat.seatType)){
                const bookedSeatExist = findSeatExist(tempArray, bookedSeat);
                if(bookedSeatExist) {
                    setExistArray(bookedSeatExist, tempArray);
                    return;
                }
            }
            tempArray[parentIndex][childIndex][availIndex] = {...tempArray[parentIndex][childIndex][availIndex], available: true};
            setNewResult(tempArray);
            return;
        }
        if(tempArray[ids[0] + 1]){
            let parentIndex = bookedSeat.seatType === 'window' ? (ids[0] === 0 ? tempArray.length-1 : 0) : (ids[0] + 1); 
            let childIndex = bookedSeat.seatType === 'window' ? (ids[0] === 0 ? ids[1] : (ids[1] + 1)) : ids[1]; 
            let availIndex = tempArray[parentIndex][childIndex].findIndex(item => item.seatType === 'aisle' && !item.booked);
            availIndex = availIndex > -1 ? availIndex : tempArray[parentIndex][childIndex].findIndex(item => item.seatType === 'window' && !item.booked);
            availIndex = availIndex > -1 ? availIndex : tempArray[parentIndex][childIndex].findIndex(item => item.seatType === 'center' && !item.booked);
            if(!tempArray[parentIndex][childIndex][availIndex] || (tempArray[parentIndex][childIndex][availIndex]['seatType'] !== bookedSeat.seatType)){
                const bookedSeatExist = findSeatExist(tempArray, bookedSeat);
                if(bookedSeatExist) {
                    setExistArray(bookedSeatExist, tempArray);
                    return;
                }
            }
            tempArray[parentIndex][childIndex][availIndex] = {...tempArray[parentIndex][childIndex][availIndex], available: true};
            setNewResult(tempArray);
            return;
        }
    }

    const bookTicket = (seat) => {
        if(seat.available) {
            const ids = (seat.id).split('');
            const passenger = passengerNo + 1;
            setPassengerNo(passenger);
            const updatedSeat = {...seat, passenger, available: false, booked: true};
            console.log("Booked passenger details", updatedSeat);
            const tempArray = JSON.parse(JSON.stringify(resultArray));
            tempArray[Number(ids[0])][Number(ids[1])][Number(ids[2])] = updatedSeat;
            setNewResult(tempArray);
            openNewAvailableSeat(tempArray, updatedSeat);
        }
    }

    return (
        <div className="App-header">
            <div className="parent" style={{gridTemplateColumns: computeColumn(resultArray)}}>
                {resultArray?.map((item1, i) => <div key={i}>
                    {item1.map((item2, j) => <div key={i+''+j} className={`child row${j + 1}`}>
                        {item2.map(item3 => <div key={item3.id} className={`${item3.seatType} ${item3.available ? 'available' : (item3.booked ? 'booked':'')}`} onClick={() => bookTicket(item3)}>
                            {item3.passenger || ''}
                        </div>)}
                    </div>)}
                </div>)}
            </div>

            <div className="character">
                <div><span className="available"/> Available</div>
                <div><span className="booked"/> Booked</div>
                <div><span className="yetToBook"/> Yet to Book</div>
            </div>
        </div>
    );
}
  
export default App;