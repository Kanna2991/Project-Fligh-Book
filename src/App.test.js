import {render, fireEvent, screen} from '@testing-library/react'
import App from './App';
import {create, act} from 'react-test-renderer';

test('Snapshot test for single column', () => {
  let renderer;

  act(() => {
    const initialRowsColumns = [ [2, 3] ];
    renderer = create(<App initialRowsColumns={initialRowsColumns}/>);
  })
  expect(renderer.toJSON()).toMatchSnapshot();
});


test('Snapshot test for two column', () => {
  let renderer;

  act(() => {
    const initialRowsColumns = [ [2, 3], [4,2] ];
    renderer = create(<App initialRowsColumns={initialRowsColumns}/>);
  })
  expect(renderer.toJSON()).toMatchSnapshot();
});

test('Snapshot test for three column', () => {
  let renderer;

  act(() => {
    const initialRowsColumns = [ [4,3], [2,3], [3,4] ];
    renderer = create(<App initialRowsColumns={initialRowsColumns}/>);
  })
  expect(renderer.toJSON()).toMatchSnapshot();
});

test('Snapshot test for four column', () => {
  let renderer;

  act(() => {
    const initialRowsColumns = [ [3,2], [4,3], [2,3], [3,4] ];
    renderer = create(<App initialRowsColumns={initialRowsColumns}/>);
  })
  expect(renderer.toJSON()).toMatchSnapshot();
});

