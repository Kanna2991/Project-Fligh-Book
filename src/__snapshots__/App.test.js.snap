// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for four column 1`] = `
<div
  className="App-header"
>
  <div
    className="parent"
    style={
      Object {
        "gridTemplateColumns": "auto auto auto auto ",
      }
    }
  >
    <div>
      <div
        className="child row1"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle available"
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
    <div>
      <div
        className="child row1"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row3"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
    <div>
      <div
        className="child row1"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row3"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
    <div>
      <div
        className="child row1"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row3"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row4"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="character"
  >
    <div>
      <span
        className="available"
      />
       Available
    </div>
    <div>
      <span
        className="booked"
      />
       Booked
    </div>
    <div>
      <span
        className="yetToBook"
      />
       Yet to Book
    </div>
  </div>
</div>
`;

exports[`Snapshot test for single column 1`] = `
<div
  className="App-header"
>
  <div
    className="parent"
    style={
      Object {
        "gridTemplateColumns": "auto ",
      }
    }
  >
    <div>
      <div
        className="child row1"
      >
        <div
          className="window available"
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row3"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="character"
  >
    <div>
      <span
        className="available"
      />
       Available
    </div>
    <div>
      <span
        className="booked"
      />
       Booked
    </div>
    <div>
      <span
        className="yetToBook"
      />
       Yet to Book
    </div>
  </div>
</div>
`;

exports[`Snapshot test for three column 1`] = `
<div
  className="App-header"
>
  <div
    className="parent"
    style={
      Object {
        "gridTemplateColumns": "auto auto auto ",
      }
    }
  >
    <div>
      <div
        className="child row1"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle available"
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row3"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
    <div>
      <div
        className="child row1"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row3"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
    <div>
      <div
        className="child row1"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row3"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row4"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="character"
  >
    <div>
      <span
        className="available"
      />
       Available
    </div>
    <div>
      <span
        className="booked"
      />
       Booked
    </div>
    <div>
      <span
        className="yetToBook"
      />
       Yet to Book
    </div>
  </div>
</div>
`;

exports[`Snapshot test for two column 1`] = `
<div
  className="App-header"
>
  <div
    className="parent"
    style={
      Object {
        "gridTemplateColumns": "auto auto ",
      }
    }
  >
    <div>
      <div
        className="child row1"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle available"
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row3"
      >
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
    <div>
      <div
        className="child row1"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
      <div
        className="child row2"
      >
        <div
          className="aisle "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="center "
          onClick={[Function]}
        >
          
        </div>
        <div
          className="window "
          onClick={[Function]}
        >
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="character"
  >
    <div>
      <span
        className="available"
      />
       Available
    </div>
    <div>
      <span
        className="booked"
      />
       Booked
    </div>
    <div>
      <span
        className="yetToBook"
      />
       Yet to Book
    </div>
  </div>
</div>
`;
